README.txt
.. contents:: Table of Contents
   :depth: 2

sc.mailchimp.api
*******************************************************************

Overview
--------

Python wrapper for the MailChimp API


Requirements
------------

    - Python >= 2.4 (http://python.org/download/)
    - simplejson >= 2.2.1 (http://pypi.python.org/pypi/simplejson)


Installation
------------

To enable this product,on a buildout based installation:

    1. Edit your buildout.cfg and add ``sc.mailchimp.api``
       to the list of eggs to install ::

        [buildout]
        ...
        eggs = 
            sc.mailchimp.api

After updating the configuration you need to run the ''bin/buildout'',
which will take care of updating your system.


Usage
-----

* Parameters

 * apikey (Your MailChimp API key eg.: 11f4654fg78g546g76876gsddsf657d3ffb-us2)
 * debug_mode (Is optional, True or False. False is default)
 * ssl (Is optional, True or False for used https protocol. True is default)


* Parameters for classe instance

 * method (More info in http://apidocs.mailchimp.com/1.3. eg.: getAccountDetails)
 * output (Output Formats for Serialized calls. json is default. eg.: json, php,xml or lolcode)
 * Others parameters (Each method requires a certain parameter, the information in the API eg.:
    mc(method='listMemberInfo',id='List X', email_address='cleber@simplesconsultoria.com.


>>> from sc.mailchimp.api import MailChimp
>>> mc = MailChimp(apikey='11f4654fg78g546g76876gsddsf657d3ffb-us2',debug_mode=True)
>>> inst = mc(method='getAccountDetails',output='json')
>>> inst


Sponsoring
----------

    * Development of this product was sponsored by:
        
        * `Simples Consultoria  <http://www.simplesconsultoria.com.br/>`_.


Credits
-------

    * Cleber J Santos (cleber at simplesconsultoria dot com dot br) - 
      Idea and implementation.


docs/HISTORY.txt
Changelog
=========

0.5 - (Unrelease)
------------------

* TODO: Merge error table with the MailChimpError method and return a dictionary with the error code and message. [cleberjsantos]

* TODO: Write usage documentation, fix Data center error setting and added headers setting [cleberjsantos]

* Error handling [cleberjsantos]

* Added new API methods [cleberjsantos]

* Initial release [cleberjsantos]


sc/mailchimp/api/errorcodes.py
#! -*- coding: UTF-8 -*-

#Return a dictionary of response codes in that shows all the response codes used by #RFC 2616.

#Table mapping response codes to messages; entries have the
#form {code: (shortmessage, longmessage)}.

responses = {
    100: ('Continue', 'Request received, please continue'),
    101: ('Switching Protocols',
          'Switching to new protocol; obey Upgrade header'),

    200: ('OK', 'Request fulfilled, document follows'),
    201: ('Created', 'Document created, URL follows'),
    202: ('Accepted',
          'Request accepted, processing continues off-line'),
    203: ('Non-Authoritative Information', 'Request fulfilled from cache'),
    204: ('No Content', 'Request fulfilled, nothing follows'),
    205: ('Reset Content', 'Clear input form for further input.'),
    206: ('Partial Content', 'Partial content follows.'),

    300: ('Multiple Choices',
          'Object has several resources -- see URI list'),
    301: ('Moved Permanently', 'Object moved permanently -- see URI list'),
    302: ('Found', 'Object moved temporarily -- see URI list'),
    303: ('See Other', 'Object moved -- see Method and URL list'),
    304: ('Not Modified',
          'Document has not changed since given time'),
    305: ('Use Proxy',
          'You must use proxy specified in Location to access this '
          'resource.'),
    307: ('Temporary Redirect',
          'Object moved temporarily -- see URI list'),

    400: ('Bad Request',
          'Bad request syntax or unsupported method'),
    401: ('Unauthorized',
          'No permission -- see authorization schemes'),
    402: ('Payment Required',
          'No payment -- see charging schemes'),
    403: ('Forbidden',
          'Request forbidden -- authorization will not help'),
    404: ('Not Found', 'Nothing matches the given URI'),
    405: ('Method Not Allowed',
          'Specified method is invalid for this server.'),
    406: ('Not Acceptable', 'URI not available in preferred format.'),
    407: ('Proxy Authentication Required', 'You must authenticate with '
          'this proxy before proceeding.'),
    408: ('Request Timeout', 'Request timed out; try again later.'),
    409: ('Conflict', 'Request conflict.'),
    410: ('Gone',
          'URI no longer exists and has been permanently removed.'),
    411: ('Length Required', 'Client must specify Content-Length.'),
    412: ('Precondition Failed', 'Precondition in headers is false.'),
    413: ('Request Entity Too Large', 'Entity is too large.'),
    414: ('Request-URI Too Long', 'URI is too long.'),
    415: ('Unsupported Media Type', 'Entity body in unsupported format.'),
    416: ('Requested Range Not Satisfiable',
          'Cannot satisfy request range.'),
    417: ('Expectation Failed',
          'Expect condition could not be satisfied.'),

    500: ('Internal Server Error', 'Server got itself in trouble'),
    501: ('Not Implemented',
          'Server does not support this operation'),
    502: ('Bad Gateway', 'Invalid responses from another server/proxy.'),
    503: ('Service Unavailable',
          'The server cannot process the request due to a high load'),
    504: ('Gateway Timeout',
          'The gateway server did not receive a timely response'),
    505: ('HTTP Version Not Supported', 'Cannot fulfill request.'),
    }


sc/mailchimp/api/__init__.py
# -*- coding: utf-8 -*-
# Cleber J Santos <cleber@simplesconsultoria.com.br>

import urllib
import urllib2
import re
from StringIO import StringIO
import errorcodes
import os

try:
    import json
except ImportError:
    try:
        import simplejson as json
    except ImportError:
        raise ImportError("Please install the simplejson package, easy_install simplejson")

class MailChimpError(Exception):
    msg = None
    code = None
    errorcodes = errorcodes.responses

    def __init__(self, code=code,msg=msg):
        self.msg = msg
        self.code = code

    def __call__(self):
        try:
            errormsg = ', '.join(self.errorcodes[int(self.code)])
        except KeyError, e:
            errormsg = self.msg
        except ValueError, e:
            errormsg = self.msg

        return {'status':self.code or 000,'msg':'Message: %s - %s' %(self.code,errormsg)}

class MailChimp(object):
    """
    MailChimp API Version 1.3
    http://apidocs.mailchimp.com/1.3/
    """

    apikey = None #myKeyFromMailChimpApi - Info: https://us2.admin.mailchimp.com/account/api/
    ssl = True # If True use https protocol else use http protocol
    debug_mode = False #Debug mode off is default
    user_agent = "sc.mailchimp.api"
    errorcodes = errorcodes.responses
    headers = {}
    api_version = '1.3'

    def __init__(self, apikey=apikey, ssl=ssl,debug_mode=debug_mode):
        try:
            self.dc = apikey.rsplit('-', 1)[-1]
        except AttributeError, e:
            merror = MailChimpError(e['code'],e['error'])
            return merror()

        self.apikey = apikey
        self.tableerrors = self.errorcodes
        self.err = None
        self.debug_mode = debug_mode
        if ssl:
            self.protocol = 'https'
        else:
            self.protocol = 'http'

        self.urlbase = "%s://%s.api.mailchimp.com/%s/" % (self.protocol, self.dc, self.api_version)

    def __call__(self, **kwargs):
        if self.apikey:
            kwargs['apikey'] = self.apikey

        if "user-agent" not in self.headers:
            self.headers['user-agent'] = self.user_agent

        data = urllib.urlencode(kwargs)
        urlfull = self.urlbase + '?' + data

        if self.debug_mode:
           print 'URL:', self.urlbase
           print 'URL Full:', urlfull
           print 'POST data:', data
           print 'API Version:', self.api_version
           print 'API Key:', self.apikey
           print 'User Agent:', self.user_agent
           print 'Headers:', self.headers

        req = urllib2.Request(urlfull,headers=self.headers or {})

        try:
            handle = urllib2.urlopen(req)
        except urllib2.HTTPError, e:
            if (e.code == 304):
                return []
            else:
               self.err = e.code
        except urllib2.URLError, e:
            if hasattr(e, 'reason'):
                self.err = {'code':000,'error':'We failed to reach a server. Reason: %s' % str(e.reason)}
            elif hasattr(e, 'code'):
                self.err = e.code

        if self.err is None:
            if handle.code == 200 and handle.msg.lower() == 'ok':
                status = 200

            resp = handle.read()
            io = StringIO(resp)
            jsReturned = json.load(io)

            if 'error' in jsReturned:
                merror = MailChimpError(jsReturned['code'],jsReturned['error']) 
                return merror()
            else:
                try:
                    jsReturned['status'] = status
                except TypeError, e:
                    pass
        else:
            if self.err.has_key('error'):
                jsReturned = MailChimpError(self.err['code'],self.err['error'])()
            else:
                jsReturned =  MailChimpError(self.err['code'])()

        return jsReturned
